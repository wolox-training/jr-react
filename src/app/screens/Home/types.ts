import { Book } from '~utils/types';

export interface BooksResponse {
  count: number;
  currentPage: number;
  nextPage: number;
  page: Book[];
  totalCount: number;
  totalPages: number;
}

export interface BooksPayload {
  limit?: number;
  page?: number;
}
