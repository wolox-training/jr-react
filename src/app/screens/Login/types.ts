import { InfoUser } from '~utils/types';

export interface LoginResponse {
  data: {
    allowPasswordChange: boolean;
    provider: string;
    uid: string;
  } & InfoUser;
}

export interface LoginBody {
  email: string;
  password: string;
}
