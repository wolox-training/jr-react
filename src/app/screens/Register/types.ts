
import { InfoUser } from '~utils/types';

export type RegisterResponse = InfoUser;

export interface RegisterBody extends Omit<InfoUser, 'id'> {
  password: string;
  passwordConfirmation: string;
}