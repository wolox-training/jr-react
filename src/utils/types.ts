import { ApiResponse } from 'apisauce';
import { Error } from '~app/hooks/useRequest';

export type Nullable<T> = T | null;

interface InfoUser{
  email: string;
  firstName: string;
  lastName: string;
  locale: string;
}

export interface UserRegister extends InfoUser{
  password: string;
  passwordConfirmation: string;
}

export interface SuccessResponse extends InfoUser{
  id: number;
}
export interface ErrorResponse {
  error: string;
  errors: {
    fullMessages: Array<string>;
  };
  status: string;
}

export type LazyRequest<D, E, P> = [Nullable<D> | unknown, boolean, Nullable<Error<E>>, (params: P) => void];

export interface ContentForm {
  onSubmit: (data: UserRegister) => void;
  request: LazyRequest<SuccessResponse, ErrorResponse, UserRegister>;
}

export type Service<Data, Error> = (data: Data) => Promise<ApiResponse<unknown, Error>>;