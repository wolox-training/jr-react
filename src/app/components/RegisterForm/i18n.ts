import i18next from 'i18next';

i18next.addResources('es', 'Register', {
  firstName: 'Nombre',
  lastName: 'Apellido',
  messageSuccess: 'Usuario registrado',
  passwordConfirmation: 'Confirmación de Contraseña'
});

i18next.addResources('en', 'Register', {
  firstName: 'Name',
  lastName: 'Surname',
  messageSuccess: 'Registered user',
  passwordConfirmation: 'Repeat Password'
});
