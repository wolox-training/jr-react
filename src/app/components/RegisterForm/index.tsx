/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';

import Input from '~components/Input';
import { emailRegex } from '~utils/inputValidations';
import { AUTH_FIELDS } from '~constants/fields';
import AlertMessage from '~components/AlertMessage';
import { ContentForm, UserRegister } from '~utils/types';
import { getErrorMessage } from '~utils/errors';

import styles from './styles.module.scss';

function RegisterForm({ onSubmit, statusApi }: ContentForm) {
  const { register, handleSubmit, errors, formState, watch } = useForm<UserRegister>({
    mode: 'all'
  });

  const validatePassword = (passwordConfirmation: string) =>
    passwordConfirmation === watch('password') ? true : 'FormValidations:passwordNotMatch';

  const [state, loading, error] = statusApi;
  return (
    <>
      <form className={styles.body} onSubmit={handleSubmit(onSubmit)}>
        {error?.problem && <AlertMessage type="error" message={getErrorMessage(error)} />}

        {state && <AlertMessage type="success" message={i18next.t('Register:messageSuccess')} />}

        <Input
          labelText={i18next.t('Register:firstName')}
          name={AUTH_FIELDS.firstName}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' }
          })}
          errorMessage={errors.firstName?.message}
        />

        <Input
          labelText={i18next.t('Register:lastName')}
          name={AUTH_FIELDS.lastName}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' }
          })}
          errorMessage={errors.lastName?.message}
        />
        <Input
          labelText={i18next.t('FormAuth:email')}
          name={AUTH_FIELDS.email}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' },
            pattern: {
              value: emailRegex,
              message: 'FormValidations:errorEmail'
            }
          })}
          type="email"
          errorMessage={errors.email?.message}
        />
        <Input
          labelText={i18next.t('FormAuth:password')}
          name={AUTH_FIELDS.password}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' }
          })}
          type="password"
          errorMessage={errors.password?.message}
        />
        <Input
          labelText={i18next.t('Register:passwordConfirmation')}
          name={AUTH_FIELDS.passwordConfirmation}
          inputRef={register({
            required: { value: true, message: 'FormValidations:errorRequired' },
            validate: validatePassword
          })}
          type="password"
          errorMessage={errors.passwordConfirmation?.message}
        />

        <button type="submit" disabled={!formState.isValid || loading} className="button btn-green">
          {i18next.t('FormAuth:btnRegister')}
        </button>
      </form>
      <div className={styles.footer}>
        <button type="button" className="button btn-white">
          {i18next.t('FormAuth:btnLogin')}
        </button>
      </div>
    </>
  );
}

export default RegisterForm;
