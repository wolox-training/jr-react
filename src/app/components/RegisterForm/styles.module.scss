@import 'variables/colors';

.body {
  margin: 15px 0;
}

.footer {
  border-top: 1px solid $dark-gray;
  padding-top: 15px;
  text-align: center;

  .language {
    margin-top: 15px;
  }
}
