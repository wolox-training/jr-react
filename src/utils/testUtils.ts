import { screen, fireEvent } from '@testing-library/react';

export const setInputByLabelText = (name: string, value: string) => {
  fireEvent.input(screen.getByLabelText(name), {
    target: {
      value
    }
  });
};
