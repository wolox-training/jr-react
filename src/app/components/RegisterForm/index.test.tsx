import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import { setInputByLabelText } from '~utils/testUtils';
import { ErrorResponse, LazyRequest, SuccessResponse, UserRegister } from '~utils/types';

import RegisterForm from './index';

const I18N_KEYS = {
  firstName: 'Register:firstName',
  lastName: 'Register:lastName',
  email: 'FormAuth:email',
  password: 'FormAuth:password',
  passwordConfirmation: 'Register:passwordConfirmation',
  buttonRegister: 'FormAuth:btnRegister'
};

const USER_TEST = {
  email: 'federico.gomez@test.com',
  firstName: 'Federico',
  lastName: 'Gomez',
  password: 'asdf1234',
  passwordConfirmation: 'asdf1234'
};

const mockOnSubmit = jest.fn();

const mockStatusApi: LazyRequest<SuccessResponse, ErrorResponse, UserRegister> = [
  null,
  false,
  null,
  value => value
];

describe('Register Form Component', () => {
  it('Renders successfully', () => {
    const { asFragment } = render(<RegisterForm onSubmit={mockOnSubmit} statusApi={[...mockStatusApi]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('If the form is not completed correctly, it is not sent', async () => {
    render(<RegisterForm onSubmit={mockOnSubmit} statusApi={[...mockStatusApi]} />);

    const buttonRegister = screen.getByRole('button', { name: I18N_KEYS.buttonRegister });

    await waitFor(() => {
      fireEvent.click(buttonRegister);
    });

    const alertErrorShow = 5;
    expect(await screen.findAllByRole('alert')).toHaveLength(alertErrorShow);
    expect(mockOnSubmit).not.toBeCalled();
  });

  test('Successful registration', async () => {
    render(<RegisterForm onSubmit={mockOnSubmit} statusApi={[...mockStatusApi]} />);
    await waitFor(() => {
      setInputByLabelText(I18N_KEYS.firstName, USER_TEST.firstName);
      setInputByLabelText(I18N_KEYS.lastName, USER_TEST.lastName);
      setInputByLabelText(I18N_KEYS.email, USER_TEST.email);
      setInputByLabelText(I18N_KEYS.password, USER_TEST.password);
      setInputByLabelText(I18N_KEYS.passwordConfirmation, USER_TEST.passwordConfirmation);
    });

    const buttonRegister = screen.getByRole('button', { name: I18N_KEYS.buttonRegister });

    await waitFor(() => {
      fireEvent.click(buttonRegister);
    });

    expect(mockOnSubmit).toHaveBeenCalled();
  });
});
