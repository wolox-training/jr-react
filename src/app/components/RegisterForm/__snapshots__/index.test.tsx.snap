// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen Register Renders successfully 1`] = `
<DocumentFragment>
  <form
    class="body"
  >
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="firstName"
      >
        Register:firstName
      </label>
      <input
        class="full-width input"
        id="firstName"
        name="firstName"
        type="text"
      />
    </div>
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="lastName"
      >
        Register:lastName
      </label>
      <input
        class="full-width input"
        id="lastName"
        name="lastName"
        type="text"
      />
    </div>
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="email"
      >
        FormAuth:email
      </label>
      <input
        class="full-width input"
        id="email"
        name="email"
        type="email"
      />
    </div>
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="password"
      >
        FormAuth:password
      </label>
      <input
        class="full-width input"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="row wrap contentInput"
    >
      <label
        class="label"
        for="passwordConfirmation"
      >
        Register:passwordConfirmation
      </label>
      <input
        class="full-width input"
        id="passwordConfirmation"
        name="passwordConfirmation"
        type="password"
      />
    </div>
    <button
      class="button btn-green"
      type="submit"
    >
      FormAuth:btnRegister
    </button>
  </form>
  <div
    class="footer"
  >
    <button
      class="button btn-white"
      type="button"
    >
      FormAuth:btnLogin
    </button>
  </div>
</DocumentFragment>
`;
